### PR TITLE
[pcre] Fix installation when only building release

### LINKED ISSUE #12566: [pcre] build failure on linux
It's making a bunch of other projects to fail to build like glib

**Host Environment**
- OS: Linux
- Compiler: gcc (Debian 8.3.0-6) 8.3.0

**To Reproduce**
`./vcpkg install pcre`

```
default@DESKTOP:/mnt/c/Users/user/Documents/code/project$ ~/code/vcpkg/vcpkg install pcre
Computing installation plan...
The following packages will be built and installed:
    pcre[core]:x64-linux
Detecting compiler hash for triplet x64-linux...
Starting package 1/1: pcre:x64-linux
Building package pcre[core]:x64-linux...
Could not locate cached archive: /home/default/.cache/vcpkg/archives/19/194f6691fd9e9619a3ca4f5ac655ff7f84f4e195.zip
-- Using cached /home/default/code/vcpkg/downloads/pcre-8.44.zip
-- Cleaning sources at /home/default/code/vcpkg/buildtrees/pcre/src/pcre-8-2e1da0dae7.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source /home/default/code/vcpkg/downloads/pcre-8.44.zip
-- Applying patch pcre-8.44_suppress_cmake_and_compiler_warnings-errors.patch
-- Using source at /home/default/code/vcpkg/buildtrees/pcre/src/pcre-8-2e1da0dae7.clean
-- Configuring x64-linux-rel
-- Building x64-linux-rel
-- Fixing pkgconfig
-- Fixing pkgconfig - release
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/lib/pkgconfig/libpcre.pc
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/lib/pkgconfig/libpcre16.pc
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/lib/pkgconfig/libpcre32.pc
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/lib/pkgconfig/libpcrecpp.pc
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/lib/pkgconfig/libpcreposix.pc
-- Fixing pkgconfig - debug
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/debug/lib/pkgconfig/libpcre.pc
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/debug/lib/pkgconfig/libpcre16.pc
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/debug/lib/pkgconfig/libpcre32.pc
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/debug/lib/pkgconfig/libpcrecpp.pc
-- Checking file: /home/default/code/vcpkg/packages/pcre_x64-linux/debug/lib/pkgconfig/libpcreposix.pc
-- CHECK_LIB_pcre_DEBUG:CHECK_LIB_pcre_DEBUG-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:239 (message):
  Library "pcre" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:348 (vcpkg_fixup_pkgconfig_check_files)
  ports/pcre/portfile.cmake:113 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:79 (include)


Error: Building package pcre:x64-linux failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: pcre:x64-linux
  Vcpkg version: 2020.06.15-unknownhash

Additionally, attach any relevant sections from the log files above.
```

### PR DESCRIPTION
**Pull request by JackBoosY**
Fixes #12566.

### CODE REVIEW

**Comment by c72578** [comment]
@JackBoosY
- There are three more conditions needed, to fix the builds when the `VCPKG_BUILD_TYPE` is set to either `release` or `debug` only. Please have a look at the following patch based upon this PR here, which handles these situations:

```diff
diff --git a/ports/pcre/portfile.cmake b/ports/pcre/portfile.cmake
index 1926975bd..7fbe0ef18 100644
--- a/ports/pcre/portfile.cmake
+++ b/ports/pcre/portfile.cmake
@@ -53,15 +53,17 @@ vcpkg_configure_cmake(
 
 vcpkg_install_cmake()
 
-foreach(FILE ${CURRENT_PACKAGES_DIR}/include/pcre.h ${CURRENT_PACKAGES_DIR}/include/pcreposix.h)
-    file(READ ${FILE} PCRE_H)
-    if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
-        string(REPLACE "defined(PCRE_STATIC)" "1" PCRE_H "${PCRE_H}")
-    else()
-        string(REPLACE "defined(PCRE_STATIC)" "0" PCRE_H "${PCRE_H}")
-    endif()
-    file(WRITE ${FILE} "${PCRE_H}")
-endforeach()
+if(NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
+    foreach(FILE ${CURRENT_PACKAGES_DIR}/include/pcre.h ${CURRENT_PACKAGES_DIR}/include/pcreposix.h)
+        file(READ ${FILE} PCRE_H)
+        if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+            string(REPLACE "defined(PCRE_STATIC)" "1" PCRE_H "${PCRE_H}")
+        else()
+            string(REPLACE "defined(PCRE_STATIC)" "0" PCRE_H "${PCRE_H}")
+        endif()
+        file(WRITE ${FILE} "${PCRE_H}")
+    endforeach()
+endif()
 
 # Create pkgconfig files
 set(PACKAGE_VERSION ${PCRE_VERSION})
@@ -74,11 +76,13 @@ if(VCPKG_TARGET_IS_LINUX)
     set(PTHREAD_CFLAGS "-pthread")
 endif()
 
-configure_file("${SOURCE_PATH}/libpcre.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcre.pc" @ONLY)
-configure_file("${SOURCE_PATH}/libpcre16.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcre16.pc" @ONLY)
-configure_file("${SOURCE_PATH}/libpcre32.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcre32.pc" @ONLY)
-configure_file("${SOURCE_PATH}/libpcrecpp.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcrecpp.pc" @ONLY)
-configure_file("${SOURCE_PATH}/libpcreposix.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcreposix.pc" @ONLY)
+if(NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
+    configure_file("${SOURCE_PATH}/libpcre.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcre.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcre16.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcre16.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcre32.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcre32.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcrecpp.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcrecpp.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcreposix.pc.in" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libpcreposix.pc" @ONLY)
+endif()
 
 # debug
 set(prefix "${CURRENT_INSTALLED_DIR}/debug")
@@ -86,11 +90,13 @@ set(exec_prefix "\${prefix}")
 set(libdir "\${prefix}/lib")
 set(includedir "\${prefix}/../include")
 
-configure_file("${SOURCE_PATH}/libpcre.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre.pc" @ONLY)
-configure_file("${SOURCE_PATH}/libpcre16.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre16.pc" @ONLY)
-configure_file("${SOURCE_PATH}/libpcre32.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre32.pc" @ONLY)
-configure_file("${SOURCE_PATH}/libpcrecpp.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcrecpp.pc" @ONLY)
-configure_file("${SOURCE_PATH}/libpcreposix.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcreposix.pc" @ONLY)
+if(NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
+    configure_file("${SOURCE_PATH}/libpcre.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcre16.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre16.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcre32.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre32.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcrecpp.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcrecpp.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcreposix.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcreposix.pc" @ONLY)
+endif()
 
 if(VCPKG_TARGET_IS_WINDOWS AND (NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug"))
     file(READ ${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre.pc LIBPCRE.PC)
```

- Open issue in case of `set(VCPKG_BUILD_TYPE debug)`:
```console
-- Performing post-build validation
The folder /include is empty or not present. This indicates the library was not correctly installed.
Found 1 error(s). Please correct the portfile:
    vcpkg\ports\pcre\portfile.cmake
```

- Triplets for testing this PR specifically under Windows are attached here
`.\vcpkg.exe install pcre:x64-windows-rel`
`.\vcpkg.exe install pcre:x64-windows-dbg`
[triplets_x64-windows-dbg-rel.zip](https://github.com/microsoft/vcpkg/files/4998821/triplets_x64-windows-dbg-rel.zip)

**Comment by JackBoosY** [comment]
@c72578 For the first modification and the second item, that is a common issue for each port. We need to discuss.

File: ports/pcre/portfile.cmake [hunk 1]
```diff
@@ -64,35 +64,36 @@ foreach(FILE ${CURRENT_PACKAGES_DIR}/include/pcre.h ${CURRENT_PACKAGES_DIR}/incl
...
-set(PACKAGE_VERSION ${PCRE_VERSION})
-set(prefix "${CURRENT_INSTALLED_DIR}")
-set(exec_prefix "\${prefix}")
-set(libdir "\${prefix}/lib")
-set(includedir "\${prefix}/include")
-if(VCPKG_TARGET_IS_LINUX)
-    # Used here in .pc.in files: Libs.private: @PTHREAD_CFLAGS@
-    set(PTHREAD_CFLAGS "-pthread")
+if (NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
+    set(PACKAGE_VERSION ${PCRE_VERSION})
```
Comment (c72578) [inline ports/pcre/portfile.cmake:68]: `set(PACKAGE_VERSION ${PCRE_VERSION})` should not be inside the `if` `"release"`

File: ports/pcre/portfile.cmake [hunk 2]
```diff
@@ -64,35 +64,36 @@ foreach(FILE ${CURRENT_PACKAGES_DIR}/include/pcre.h ${CURRENT_PACKAGES_DIR}/incl
...
+    set(prefix "${CURRENT_INSTALLED_DIR}")
+    set(exec_prefix "\${prefix}")
+    set(libdir "\${prefix}/lib")
+    set(includedir "\${prefix}/include")
+    if(VCPKG_TARGET_IS_LINUX)
+        # Used here in .pc.in files: Libs.private: @PTHREAD_CFLAGS@
+        set(PTHREAD_CFLAGS "-pthread")
+    endif()
+
```
Comment (c72578) [inline ports/pcre/portfile.cmake:77]: Here is the previously suggested starting point for:
`if (NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")`

File: ports/pcre/portfile.cmake [hunk 3]
```diff
@@ -64,35 +64,36 @@ foreach(FILE ${CURRENT_PACKAGES_DIR}/include/pcre.h ${CURRENT_PACKAGES_DIR}/incl
...
+    set(libdir "\${prefix}/lib")
+    set(includedir "\${prefix}/../include")
+    
+    configure_file("${SOURCE_PATH}/libpcre.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcre16.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre16.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcre32.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre32.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcrecpp.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcrecpp.pc" @ONLY)
+    configure_file("${SOURCE_PATH}/libpcreposix.pc.in" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcreposix.pc" @ONLY)
 
-if(VCPKG_TARGET_IS_WINDOWS)
```
Comment (c72578) [inline ports/pcre/portfile.cmake:95]: Please do not remove this line, the `POSTFIX` is Windows specific.
https://vcs.pcre.org/pcre2/code/trunk/CMakeLists.txt?revision=1265&view=markup

```cmake
IF(WIN32)
  SET(CMAKE_DEBUG_POSTFIX "d")
ENDIF(WIN32)
```

**Comment by c72578** [comment]
@JackBoosY Could you please rename this PR and replace `debug` with `release` in the title:
[pcre] Fix installation when only building debug
->
[pcre] Fix installation when only building release
This is the main goal of the PR and fixes the issue as described in #12566, where `set(VCPKG_BUILD_TYPE release)` was used.

File: ports/pcre/portfile.cmake [hunk 4]
```diff
@@ -73,41 +73,43 @@ if(VCPKG_TARGET_IS_LINUX)
...
+            "-lpcre" "-lpcred"
+        )
+        vcpkg_replace_string(${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre16.pc
+            "-lpcre16" "-lpcre16d"
+        )
+        vcpkg_replace_string(${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcre32.pc
+            "-lpcre32" "-lpcre32d"
+        )
+        vcpkg_replace_string(${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libpcrecpp.pc
+            "-lpcred" "-lpcrecppd"
```
Comment (JackBoosY) [inline ports/pcre/portfile.cmake:107]:
```suggestion
            "-lpcre -lpcrecpp" "-lpcred -lpcrecppd"
```

**Review by NancyLi1013 — APPROVED** [review]

**Pull request merged**